docs/proposals/2019-02-08 - Fallback Content Elements.md

```diff
@@ -0,0 +1,90 @@
+
+## Alternatives
+
+Each `raw_html` element should have an `alternative` and at least a `basic` array with non-`raw_html` content. An example of the default behaviour:
+
+```
+{
+  "type": "raw_html",
+  "content": "<script src=\"https://www.washingtonpost.com/stat/cool.js\"></script>",
+  "alternative": [
+    {
+      "type": "text",
+      "content": "<b>This content is unavailable on this platform</b>"
+    }
+  ]
+}
+```
+
+# Example 1
+
+The graphics team has created a JS rendered graphic that scales to browser viewport size. They also generate static `.png`s as a fallback for mobile and tablet.
+
+```
+{
+  "type": "raw_html",
+  "content": "<div class=\"graphics\"><!-- Graphics HTML and JS --></div>",
+  "alternative": [
+    {
+      "type": "image",
+      "url": "https://www.coolpublishing.com/graphics/alt-cool-1.png"
+    }
+  ]
+}
+```
+
+# Example 2 (future implementation of multiple elements)
+
+The social team is trying to build awareness of the Twitch channel so they are embedding the Twitch stream where normally we would use the Goldfish live video. They provide the normal video as an alternative as well as a note to encourage users to open the twitch channel.
+
+```
+{
+  "type": "raw_html",
+  "content": "<iframe src=\"https://player.twitch.tv/?channel=coolpublishing\" frameborder=\"0\" allowfullscreen=\"true\" scrolling=\"no\" height=\"378\" width=\"620\"></iframe><a href=\"https://www.twitch.tv/coolpublishing?tt_content=text_link&tt_medium=live_embed\" style=\"padding:2px 0px 4px; display:block; width:345px; font-weight:normal; font-size:10px; text-decoration:underline;\">Watch live video from coolpublishing on www.twitch.tv</a>",
+  "alternative": [
+    {
+      "type": "video",
+      // PostTV, etc ...
+    },
+    {
+      "type": "text",
+      "content": "<b>We are also streaming live on <a href=\"https://www.twitch.tv/coolpublishing\">Twitch</a></b>"
+    }
+  ]
+}
+```
+
+# Concerns
+
+## Wouldn't this be better flipped? Where the `raw_html` was available as an alternative to an content_element? 
+
+Yes, honestly, in my opinion that is the "right" way to handle `raw_html`. The `raw_html` element should ONLY be allowed as an alternative to "pure" ANS content elements and in a well-known named one (ex `browser`) to help rendering applications make the right decision on content. The `raw_html` element is an escape hatch to ANS and defeats many of the benefits of ANS to the rendering layer. There should never be `raw_html` at the root level of the content element list. What I am calling the `basic` alternative above should be the content element or collection in an `element_group` and the `raw_html` could be in the alternatives for those applications that can use it. 
```

Comment:
This assumes a lot about the ability of adapters to sort through, say, embed code, and build an appropriate set of alternatives.  I mean, adapters would have to know all the things.  

In my opinion, raw_html is a perfectly acceptable root element.  It doesn't always contain "bad" elements, just stuff that it is maybe not worth extracting.

Reply:
This seems to assume that HTML is always renderable, or that we are dealing with HTML-native clients. This is absolutely not the case. `raw_html` is explicitly placing client-side display logic for one particular client directly into the document. This is obviously accepted in some cases due to lack of better options, but IMHO it shouldn’t be looked at as a valid base level element, at least not without a fallback/default option.